File: frontend/src/api/modelCredentials.ts
import api from './client';

export type ModelProvider = 'openai' | 'anthropic' | 'openrouter' | 'ollama';

/** Providers that can back voice input (mirrors backend SpeechProvider). */
export type SpeechProvider = Extract<ModelProvider, 'openai'>;
export const SPEECH_CAPABLE_PROVIDERS: readonly SpeechProvider[] = ['openai'];

export function isSpeechProvider(
  provider: ModelProvider | null | undefined,
): provider is SpeechProvider {
  return !!provider && (SPEECH_CAPABLE_PROVIDERS as readonly string[]).includes(provider);
}

export interface ModelCredential {
  provider: ModelProvider;
  model: string;
  light_provider?: ModelProvider | null;
  light_model?: string | null;
  heavy_provider?: ModelProvider | null;
  heavy_model?: string | null;
  vision_provider?: ModelProvider | null;
  vision_model?: string | null;
  speech_provider?: ModelProvider | null;
  speech_model?: string | null;
  key_fingerprint: string;
  light_key_fingerprint?: string | null;
  heavy_key_fingerprint?: string | null;
  vision_key_fingerprint?: string | null;
  speech_key_fingerprint?: string | null;
  is_active: boolean;
  validated_at?: string | null;
  last_used_at?: string | null;
  updated_at?: string | null;
}

export interface ModelCredentialUpsertBody {
  provider: ModelProvider;
  model: string;
  /** Required on first create; omit on update to keep stored key. */
  api_key?: string;
  light_model?: string;
  light_provider?: ModelProvider;
  light_api_key?: string;
  heavy_model?: string;
  heavy_provider?: ModelProvider;
  heavy_api_key?: string;
  vision_model?: string;
  vision_provider?: ModelProvider;
  vision_api_key?: string;
  /** Voice input. Empty string turns the slot off. */
  speech_model?: string;
  speech_provider?: ModelProvider;
  speech_api_key?: string;
}

export interface ModelCredentialValidateBody {
  provider: ModelProvider;
  api_key: string;
}

export const modelCredentialsApi = {
  async get(): Promise<ModelCredential | null> {
    const res = await api.get<{ credential: ModelCredential | null }>(
      '/users/me/model-credentials',
    );
    return res.data.credential;
  },

  async upsert(body: ModelCredentialUpsertBody): Promise<ModelCredential> {
    const res = await api.post<{ credential: ModelCredential }>(
      '/users/me/model-credentials',
      body,
    );
    return res.data.credential;
  },

  async revoke(): Promise<void> {
    await api.delete('/users/me/model-credentials');
  },

  async validate(
    body: ModelCredentialValidateBody,
  ): Promise<{ valid: boolean; message: string }> {
    const res = await api.post<{ valid: boolean; message: string }>(
      '/users/me/model-credentials/validate',
      body,
    );
    return res.data;
  },
};

export const PROVIDER_CONSOLE_URLS: Record<ModelProvider, string> = {
  openai: 'https://platform.openai.com/api-keys',
  anthropic: 'https://console.anthropic.com/settings/keys',
  openrouter: 'https://openrouter.ai/keys',
  ollama: 'https://ollama.com/settings/keys',
};

export const PROVIDER_LABELS: Record<ModelProvider, string> = {
  openai: 'OpenAI',
  anthropic: 'Anthropic',
  openrouter: 'OpenRouter',
  ollama: 'Ollama Cloud',
};

/** User-facing model role labels (AI Models settings). */
export const MODEL_SLOT_LABELS = {
  default: 'Primary',
  light: 'Quick replies',
  heavy: 'Complex tasks',
  vision: 'Images',
  speech: 'Voice input',
} as const;

export const MODEL_SLOT_HINTS = {
  default: 'Your main model for everyday chat and most answers.',
  light: 'A faster model for simple questions and short back-and-forth.',
  heavy: 'A stronger model when something needs more thought or several steps.',
  vision: 'For photos, screenshots, and other uploads with visuals.',
  speech:
    'Turns speech into text in the chat box. Everyone in workspaces you own dictates on this key.',
} as const;

/** Plain-language labels for workspace key policy (from agent_key_mode). */
export const AGENT_KEY_MODE_LABELS: Record<string, string> = {
  hybrid: 'Your keys when set, otherwise shared',
  byo_strict: 'Your keys required',
  platform_only: 'Shared keys only',
};

File: backend/app/schemas/model_credentials.py
"""Schemas for per-user model credential (BYOK) API — slot-based."""

from __future__ import annotations

from typing import FrozenSet, Literal, Optional, get_args

from pydantic import BaseModel, Field, model_validator

ModelProvider = Literal["openai", "anthropic", "openrouter", "ollama"]

# Providers that can back the ``speech`` (voice input) slot. A subset of
# ``ModelProvider``: only vendors with a streaming speech-to-text API qualify.
SpeechProvider = Literal["openai"]
SPEECH_CAPABLE_PROVIDERS: FrozenSet[str] = frozenset(get_args(SpeechProvider))


class ModelCredentialUpsertRequest(BaseModel):
    """Slot-based BYOK upsert. ``default`` (provider+model) is required.

    ``api_key`` is required on first create; omit or send empty on update to keep
    the stored default key.
    """

    provider: ModelProvider
    model: str = Field(..., min_length=1, max_length=128)
    api_key: Optional[str] = Field(default=None, max_length=512)

    light_model: Optional[str] = Field(default=None, max_length=128)
    light_provider: Optional[ModelProvider] = None
    light_api_key: Optional[str] = Field(default=None, min_length=8, max_length=512)

    heavy_model: Optional[str] = Field(default=None, max_length=128)
    heavy_provider: Optional[ModelProvider] = None
    heavy_api_key: Optional[str] = Field(default=None, min_length=8, max_length=512)

    vision_model: Optional[str] = Field(default=None, max_length=128)
    vision_provider: Optional[ModelProvider] = None
    vision_api_key: Optional[str] = Field(default=None, min_length=8, max_length=512)

    # Speech-to-text (voice input) slot. Opt-in: an empty ``speech_model`` means off.
    speech_model: Optional[str] = Field(default=None, max_length=128)
    speech_provider: Optional[ModelProvider] = None
    speech_api_key: Optional[str] = Field(default=None, min_length=8, max_length=512)

    @model_validator(mode="after")
    def _speech_provider_is_capable(self) -> "ModelCredentialUpsertRequest":
        if not (self.speech_model or "").strip():
            return self
        effective = self.speech_provider or self.provider
        if effective not in SPEECH_CAPABLE_PROVIDERS:
            capable = ", ".join(sorted(SPEECH_CAPABLE_PROVIDERS))
            raise ValueError(
                f"voice input needs a speech-to-text provider ({capable}); "
                "set speech_provider"
            )
        return self

    @model_validator(mode="after")
    def _dual_provider_requires_keys(self) -> "ModelCredentialUpsertRequest":
        pairs = (
            ("light", self.light_provider, self.light_api_key),
            ("heavy", self.heavy_provider, self.heavy_api_key),
            ("vision", self.vision_provider, self.vision_api_key),
            ("speech", self.speech_provider, self.speech_api_key),
        )
        for label, alt_provider, alt_key in pairs:
            if not alt_provider or alt_provider == self.provider:
                continue
            if not (alt_key or "").strip():
                raise ValueError(
                    f"{label}_api_key is required when {label}_provider "
                    "differs from provider"
                )
        return self


class ModelCredentialResponse(BaseModel):
    provider: ModelProvider
    model: str
    light_provider: Optional[ModelProvider] = None
    light_model: Optional[str] = None
    heavy_provider: Optional[ModelProvider] = None
    heavy_model: Optional[str] = None
    vision_provider: Optional[ModelProvider] = None
    vision_model: Optional[str] = None
    speech_provider: Optional[ModelProvider] = None
    speech_model: Optional[str] = None
    key_fingerprint: str
    light_key_fingerprint: Optional[str] = None
    heavy_key_fingerprint: Optional[str] = None
    vision_key_fingerprint: Optional[str] = None
    speech_key_fingerprint: Optional[str] = None
    is_active: bool = True
    validated_at: Optional[str] = None
    last_used_at: Optional[str] = None
    updated_at: Optional[str] = None


class ModelCredentialValidateRequest(BaseModel):
    provider: ModelProvider
    api_key: str = Field(..., min_length=8, max_length=512)


class ModelCredentialValidateResponse(BaseModel):
    valid: bool
    message: str = ""
